enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,341 @@
+---
+title: oc-authorization-grant-login
+authors:
+  - "@arjunrn"
+reviewers:
+  - "@stlaz"
+  - "@s-urbaniak"
+approvers:
+  - "@soltysh"
+  - "@mfojtik"
+creation-date: 2021-11-08
+last-updated: 2021-11-08
+status: provisional
+---
+
+# oc Authorization Grant Login
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The current CLI login process involves multiple interactive steps where the user has to enter credentials multiple
+times. This process can be simplified so that the `oc` tool can be used to login directly without having to first login
+to the cluster console, the OAuth server and finally copying the login command. The `oc` tool can be modified to act as
+a public client to the Openshift OAuth server and received an access token
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified by using OAuth Authorization Grants for getting the token in `oc`. This way the user has
+to perform a login only once and the `oc` tool is configured automatically without any user input.
+
+### Goals
+
+Provide a single-step login for identity providers that don't post HTTP challenges.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### API Extensions
+
+None
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have a new flag `--auto` which indicates to `oc` that it should use the Oauth2 Authorization
```

Comment:
IIUC making this the default would change existing behavior. The existing behavior is to call the authorization endpoint and set the redirect URI to the implicit token endpoint and if there is an identity provider which supports challenge based authentication it takes the username and password passed along with the `login` command and performs authorization. @stlaz please correct me if I'm wrong.
That said it might be possible for the server to indicate how to proceed and then use that to determine behavior. Let me investigate further.

Reply:
That's why I asked for alternatives and their pros & cons, so that we know in the future what was considered and why we did what we did :)

Reply:
I"ve explained why the new flow cannot be the default behavior and why a new flag is required.

Reply:
There already is default behavior in `oc` even when no credentials are provided I don't think it's possible to have the new login flow as default in any scenario.

Reply:
> It'd be nice to be able to default to this in case there are no HTTP challenges/redirects from the oauth-server, instead of printing a message for the users to go to login to the oauth-server

What about the inverse case. When there are identity providers which support challenges but the user wants to log in with the browser because their identity is present in an identity provider which doesn't support challenges. This would would mean a flag is required. So if we support the fallback case and the flag case that means there are 2 ways launch a browser for login. And one if them is conditional. I'm not sure that is the best user experience.

Reply:
I'm not sure if `--browser` is the most suitable name, but I can't think of anything better yet, so let's go with that for now. If I come up with something better, we can always change that during implementation.